### ISSUE TITLE
Server ended up doing state group resolution on pretty much every CS request.

### ISSUE THREAD

**Issue by ara4n** [matrix-org/synapse#1761]
Whilst investigating #1748, i saw even weirder behaviour in @vberger's logs.  As well as having accumulated a bunch of graph extremities for #rust (see #1760), something was causing synapse to try to do resolution on current room state for #rust on pretty much every request (both CS and SS!)... as well as hundreds which weren't even associated with a request context.

for instance, an innocent 'set present' CS request would take 30s, with 20s of CPU time, thanks to the merge:

```
2017-01-04 15:23:35,238 - synapse.access.http.8008 - 59 - INFO - PUT-273116 - 213.137.178.134 - 8008 - Received request: PUT /_matrix/client/r0/presence/%40levans%3Asafaradeg.net/status?access_token=<redacted>
2017-01-04 15:23:35,272 - synapse.storage._base - 242 - WARNING - PUT-273116 - [TXN OPERROR] {update_presence-fa365} could not serialize access due to concurrent update
2017-01-04 15:23:35,754 - synapse.state - 310 - INFO - PUT-273116 - Resolving state for !PcCidzjUKXHKImhTrB:matrix.org with 21 groups
2017-01-04 15:23:35,945 - synapse.state - 325 - INFO - PUT-273116 - Resolving conflicted state for '!PcCidzjUKXHKImhTrB:matrix.org'
2017-01-04 15:24:04,359 - synapse.access.http.8008 - 91 - INFO - PUT-273116 - 213.137.178.134 - 8008 - {@levans:safaradeg.net} Processed request: 29121ms (19259ms, 60ms) (53ms/1) 22B 200 "PUT /_matrix/client/r0/presence/%40levans%3Asafaradeg.net/status?access_token=<redacted> HTTP/1.0" "Mozilla/5.0 (X11; Linux x86_64; rv:50.0) Gecko/20100101 Firefox/50.0"
```

...similarly /sync requests, and /typing, etc.  wtf?

**Comment by ara4n** [comment]
More logging for resolving state groups was added in #1767 which will hopefully help explain this

**Comment by ara4n** [comment]
presumably related to #1760

**Comment by richvdh** [comment]
we have to figure out what the current state of a room is before we let you send an event (including typing notifications); likewise when you request the state of that room. To calculate current state, we have to state-resolve all of the forward extremities.

I wouldn't necessarily expect us to have to do state resolution on an incremental sync unless there's been a change in that room (which, given !PcC is quite busy, would happen quite often). 

I *would* expect us to hit caches more often than not when doing the state resolution, so it should be pretty quick.

In short, nothing here jumps out at me.

**Comment by ara4n** [comment]
sounds like this was behaving as expected then

**Comment by ara4n** [comment]
Further clarification on why /sync does state res:

 * [Matthew] (although i never understood why /sync has to do state res - shouldn't we only ever state res on receiving on fed)?
 * [Erik] sync has weird state semantics
in that its not the latest state
so we go back and try and figure out what the state was at the start of the timeline
 * [Matthew] okay, although why would that need state res? shouldn't we have resolved state already when we received stuff over fed?
 * [Erik] Because the start of the timeline might be "during" a fork in the DAG
 * [Matthew] but wouldn't we have figured that out at the time?
 * [Matthew] or am i just fundamentally misunderstanding if/where we cache state res results?
 * [Erik] We only store the state for the current extremities (i.e. current state) and for state at particular events
modulo some in memory caching
 * [Matthew] ah, ok. which particular events in particular?
 * [richvdh] all events
but the problem is that sync's state doesn't correspond to a particular point on the dag
(aiui)
 * [Erik] yup
